WordPress example major update.

jeffmendoza opened the pull request:
Rewrite the WordPress example to use deployments, persistent volume claims, and update the sample flow to the latest guidelines.

This PR goes along with #21717. Comments here will be used to shape the example guidelines

cc @bgrant0607 @kubernetes/examples

jeffmendoza committed:
WordPress example major update.

Rewrite the WordPress example to use deployments, persistent volume
claims, and update the sample flow to the latest guidelines.

jeffmendoza commented:
After updating the guidelines, I need to do a few updates to this example to match:
* TOC
* Quickstart
* Remove prompts from commands before example output, change from `console` to `shell` type and move the output to a separate block.

bgrant0607 commented:
cc @janetkuo @pwittrock

jeffmendoza commented:
Updated with TOC, quickstart, and shell/console output.

bgrant0607 commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,231 @@ Documentation for other releases can be found at
+[Persistent Volumes](../../docs/user-guide/persistent-volumes.md) and
+[Secrets](../../docs/user-guide/secrets.md).
+
+We'll use the [mysql](https://registry.hub.docker.com/_/mysql/) and
```
Is there (or could there be) an image on GCR? Docker hub isn't very reliable.
https://github.com/kubernetes/kubernetes/issues/15251

jeffmendoza replied:
Firstly, all custom built images should be on gcr.

I think we should use official images where we can, this reduces our maintenance, and is best for the user consuming the example. They will know this is a community developed, accepted, and maintained image they can rely on to begin building upon the example.

As for e2e, I agree it should not pull from Docker Hub, Maybe we can sed out the Hub images for our own for e2e?

bgrant0607 commented:
```
FAIL: changes needed but not made due to --verify
/workspace/kubernetes/examples/ is out of date. Please run hack/update-generated-docs.sh
```

We may need to change the examples_test also.

bgrant0607 commented:
Oh, you updated the examples_test, but I didn't think it worked for multi-resource files yet.

bgrant0607 commented on examples/mysql-wordpress-pd/README.md:
````diff
@@ -34,301 +34,231 @@ Documentation for other releases can be found at
-```
+Host paths are volumes mapped to directories on the host. These should
+be used for testing or single-node clusters only, as the data will not
+be moved between nodes if the pod is recreated on a new node. Create
````
The cassandra example shows a DaemonSet, which is another option.

saad-ali replied:
The DeamonSet for Cassandra is pretty cool--how difficult is it to set up a distributed MySQL deployment?

> These should be used for testing or single-node clusters only, as the data will not be moved between nodes if the pod is recreated on a new node. 

I would emphasis  this more (italics maybe), maybe clarify what exactly happens as a result (MySQL pod gets moved to a new node, data remains on old node, apparent data loss). And maybe consider using a nodeName or nodeSelector to ensure the pod always lands on the same machine? Still bad if the node goes down, but at least data won't disappear due to pod getting rescheduled.

bgrant0607 commented:
cc @pwittrock 

cc @saad-ali re. use of PVs

I'll get back to this after the 1.2 branch cut.

pwittrock commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,231 @@ Documentation for other releases can be found at
+[Services](../../docs/user-guide/services.md) to front each
+deployment.
+
+This example demonstrates several useful things, including: how to set
```
A lot of information in this paragraph.  What do you think of a list / bullet point structure mapping k8s abstraction to what it provides in this example.  As an example:

#### Demonstrated Concepts:
- use `Volumes` to define persistent disks (disk lifecycle not tied to the *Pods*)
- use `Services` to enable *Pods* to locate one another
- use `External Load Balancers` to expose *Services* externally

jeffmendoza replied:
Updated to use a list, and separate out the Kubernetes concepts.

saad-ali commented on examples/mysql-wordpress-pd/README.md:
````diff
@@ -34,301 +34,231 @@ Documentation for other releases can be found at
+
+```shell
+gcloud compute disks create --size=20GB --zone=<zone> wordpress-1
+gcloud compute disks create --size=20GB --zone=<zone> wordpress-2
````
The purpose of creating two PVs here is to demonstrate that the PVC binder will choose one at random?

jeffmendoza replied:
One is used by MySQL, the other by WordPress.

pwittrock commented on examples/mysql-wordpress-pd/README.md:
````diff
@@ -34,301 +34,231 @@ Documentation for other releases can be found at
+trailing newline.
+
+```shell
+kubectl create -f https://raw.githubusercontent.com/kubernetes/kubernetes/master/examples/mysql-wordpress-pd/local-volumes.yaml
````
Even though it should be obvious what each line does from the yaml file name, the url is so long that I need to scroll over to see the file name.  I find the length of the url adds a bit of cognitive work.

Consider either creating an environment variable for `https://raw.githubusercontent.com/kubernetes/kubernetes/master/examples/mysql-wordpress-pd` to shorten each command to be more parseable, or add a comment before each line.

jeffmendoza replied:
Updated to use an env var. Not the Quickstart though. I don't think it makes sense there, as the quickstart is supposed to be the shortest number of commands, something you c&p without caring.

Had to update the tests `cmd/mungedocs/kubectl_dash_f.go` PTAL

bgrant0607 replied:
The following also works:
https://github.com/kubernetes/kubernetes/raw/master/examples/mysql-wordpress-pd/wordpress-service.yaml

bgrant0607 replied:
BTW, did you confirm that `kubectl create -f` works with https?

bgrant0607 replied:
Maybe we can create a redirector, similar to issues.k8s.io?

bgrant0607 replied:
code.k8s.io works, but we need a raw.k8s.io

bgrant0607 replied:
?raw=true doesn't survive redirect, sadly. May be fixable

saad-ali commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,231 @@ Documentation for other releases can be found at
+[GCE getting started guide](../../docs/getting-started-guides/gce.md)
+for full details and other options for starting a cluster.
+
+Create two persistent disks. You will need to create the disks in the
```
k8s v1.2 adds support dynamic provisioning support (https://docs.openshift.org/latest/install_config/persistent_storage/dynamically_provisioning_pvs.html#dynamic-pvs-requesting-storage), so that volumes don't need to be manually provisioned ahead of time--but it's still experimental. After v1.3 is launched, we should update this example to use dynamic provisioning.

jeffmendoza replied:
Sounds good.

pwittrock commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,231 @@ Documentation for other releases can be found at
+modular. Not all clusters are the same. These are the requirements for
+this example.
+
+* Kubernetes version 1.2 is required due to using newer features, such
```
How can the user easily test that their cluster meets each of these requirements?

jeffmendoza replied:
The version is easy with `kubectl version`. For the others, examples should link out to relevant documentation.

bgrant0607 replied:
Link to 1.2? And/or a document that explains how to check?

http://releases.k8s.io
https://github.com/kubernetes/kubernetes/blob/master/docs/user-guide/kubectl/kubectl_version.md

pwittrock replied:
Rereading this, I think this is pretty confusing and we need to make this section much simpler and assume less of the user.  The links don't clearly state how to check if these are enabled for a given cluster.  The _Persistent Volume Claims_ links to a document self described as `This document describes the current state of PersistentVolumes in Kubernetes. Familiarity with volumes is suggested.`  Reading and understanding the PV doc shouldn't be required for me to do the example. 

This should probably be integrated into setting up the cluster.

saad-ali commented:
PV/PVC stuff LGTM mod some minor suggestions.

pwittrock commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,231 @@ Documentation for other releases can be found at
+## Cluster Requirements
+
+Consult a
+[Getting Started Guide](../../docs/getting-started-guides/)
```
Not sure if this is only due to the PR, but when I click on this link a list of a dozen .md files shows up.  This is a bit of a jarring experience.  Which setup they pick will have implications later on that are not entirely clear at this point.  (GCE PD requires a GCE cluster).  Which should 2 sane defaults for the 2 possible paths (GCE PD and HostPath PD), and provide the full list of possible hosting options as optional.  We should also make it clear which hosting solutions support GCE PD.

jeffmendoza replied:
I think the getting-started-guides README needs an overhaul to highlight certain deployments. I'll update this link to link directly to the README, so there isn't the whole directory listing to scroll past.

For the examples, we don't want to have cluster start up instructions in each one, this should be covered by this link. Ideally each example lists the requirements, and doesn't cover all the possible environments that will work. I think I'll change this to list the requirements first, and then link to the getting-started-guides after.

For examples that use PVs, we want to change them all to use PVCs, and then add a requirement that a PV be available to be claimed. Thus removing any specific cloud provider dependencies.

Maybe I'll take out both the GCE and Host Path sections, and only link to here: https://github.com/kubernetes/kubernetes/tree/master/docs/user-guide/persistent-volumes where we will need to have these documented.

pwittrock commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,231 @@ Documentation for other releases can be found at
+### GCE Persistent Disk
+
+Please see the
+[GCE getting started guide](../../docs/getting-started-guides/gce.md)
```
What is the relationship between this and the section describing settings up a cluster under `Cluster Requirements`?  This link leads to another separate example:
>
The example below creates a Kubernetes cluster with 4 worker node Virtual Machines and a master Virtual Machine (i.e. 5 VMs in your cluster).

Bootstrapping the user into a cluster is a real challenge for our examples because it is quite jarring to context switch from one example (setting up wordpress) to another example (starting a k8s cluster), and puts a cognitive burden on the reader.

We should probably merge this into the setting up the cluster section and just add a note hear stating that this section requires you chose the GCE option in that section.

jeffmendoza replied:
Yep, I'll remove the directive "Please see the ..." and reword.

pwittrock commented on examples/mysql-wordpress-pd/README.md:
````diff
@@ -34,301 +34,231 @@ Documentation for other releases can be found at
-```sh
-$ kubectl get pods
+```shell
+kubectl create secret generic mysql-pass --from-file=password.txt
````
Explain this command.  What are the `generic` and `mysql-pass` arguments defining?

pwittrock replied:
Briefly explain how this becomes visible to the pods (where it is defined in the deployment and where the containers see it)

pwittrock replied:
Ah, this is explained later.

jeffmendoza replied:
Added some explanation below the command.

pwittrock commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,231 @@ Documentation for other releases can be found at
 
-So if we label our Kubernetes mysql service `mysql`, the wordpress pod will be able to use the Docker-links-compatible environment variables, defined by Kubernetes, to connect to the database.
+Also note that we specified the password by referencing the secret
+`mysql-pass` that we created above. Secrets can have multiple
```
Would be helpful to explicitly state where / how it was referenced - in the mysql-deployment.yaml env section.

May something like..."The password is provided to the mysql container as an environment variable called `MYSQL_ROOT_PASSWORD`.  This is defined in the `env` section in mysql-deployment.yaml"

jeffmendoza replied:
Reworded

pwittrock commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,231 @@ Documentation for other releases can be found at
-The [`mysql-service.yaml`](mysql-service.yaml) file looks like this:
+It may take a short period before the new pod reaches the `Running`
+state.  List all pods to see the status of this new pod and the
+cluster node that it is running on:
```
Is the state of the node exposed through the provided command?

jeffmendoza replied:
No, removed.

pwittrock commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,231 @@ Documentation for other releases can be found at
 
-[Download example](wordpress.yaml?raw=true)
-<!-- END MUNGE: EXAMPLE wordpress.yaml -->
+You can take a look at the logs for a pod by using `kubectl log`. Copy
```
nit: state that the pod logs contain the stderr + stdout of running containers

jeffmendoza replied:
Fixed.

pwittrock commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,231 @@ Documentation for other releases can be found at
-<!-- END MUNGE: EXAMPLE wordpress-service.yaml -->
+Also in [mysql-deployment.yaml](mysql-deployment.yaml) we created a
+service to allow other pods to reach this mysql instance. The name is
+`wordpress-mysql` which is resolvable to the pod IP.
```
nit: s/which is resolvable/resolves/

jeffmendoza replied:
Fixed.

pwittrock commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,231 @@ Documentation for other releases can be found at
+set up the wordpress service behind an external IP.
 
-or by listing the forwarding rules for your project:
+Find the external IP for your WordPress service. It may take a minuite
```
s/minuite/minute/

pwittrock replied:
nit: clarify where the external IP is assigned (to have an external IP assigned to the service)

jeffmendoza replied:
Fixed.

pwittrock commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -336,27 +266,33 @@ http://<external-ip>
 
-If you are just experimenting, you can take down and bring up only the pods:
+All of the resources are labeled with `app=wordpress`, so you can
+easily bring them down with:
```
nit: may "bring them down using a label selector"

jeffmendoza replied:
Fixed.

pwittrock commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -336,27 +266,33 @@ http://<external-ip>
-When you restart the pods again (using the `create` operation as described above), their services will pick up the new pods based on their labels.
+Later, re-creating the resources with the original commands will pick
+up the original disks with all your data intact. We did not delete the
+PV Claims. If we did that, our disks might be claimed by other pods in
```
nit: This might be easier to read as "Because we did not delete the PV Claims, no other pods in the cluster could claim them after we deleted our pods.  Keeping the PV Claims also ensured recreating the Pods did not cause the PD to switch Pods."

jeffmendoza replied:
Fixed.

pwittrock commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,231 @@ Documentation for other releases can be found at
+for persistent storage and a secret for the password.
 
-Then, find the external IP for your WordPress service by running:
+The [WordPress image](https://hub.docker.com/_/wordpress/) can accept
```
nit: a bit awkward.  Consider breaking up the sentence and dropping the conjunction.  Consider making the sentence more active + concise through s/can accept/accepts/ 

also s/databse/database/

The WordPress image accepts the database hostname through the environment variable `WORDPRESS_DB_HOST`.  We set the env value to the name of the MySQL service we created:  `wordpress-mysq`.

jeffmendoza replied:
Fixed.

pwittrock commented on examples/mysql-wordpress-pd/wordpress-deployment.yaml:
```diff
@@ -0,0 +1,61 @@
+        tier: frontend
+    spec:
+      containers:
+      - image: wordpress:latest
```
do we recommend using the latest tag, or a stable version?

jeffmendoza replied:
Fixed.

bgrant0607 replied:
stable

bgrant0607 commented:
For others, our new example guidelines:
https://github.com/kubernetes/kubernetes/blob/master/examples/guidelines.md

pwittrock commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,231 @@ Documentation for other releases can be found at
-
-We'll define and start a [service](../../docs/user-guide/services.md) that lets other pods access the mysql database on a known port and host.
-We will specifically name the service `mysql`.  This will let us leverage the support for [Docker-links-compatible](../../docs/user-guide/services.md#how-do-they-work) service environment variables when we set up the wordpress pod. The wordpress Docker image expects to be linked to a mysql container named `mysql`, as you can see in the "How to use this image" section on the wordpress docker hub [ ...
+Note that we've defined a volume mount for `/var/lib/mysql`, and then
```
s/Note/Above/

jeffmendoza replied:
Reworded.

pwittrock commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,231 @@ Documentation for other releases can be found at
 
-#### Check the running pod on the Compute instance
+Now that the persistent disks and secrets are defined, the Kubernetes
+pods can be launched. We'll start with MySQL using
```
nit: can be more concise:

"Start MySQL using..."

jeffmendoza replied:
Fixed.

pwittrock commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,231 @@ Documentation for other releases can be found at
 
-Now, we can visit the running WordPress app.
-Use the external IP that you obtained above, and visit it on port 80:
+Now, we can visit the running WordPress app. Use the external IP that
```
...external IP of the service that...

jeffmendoza replied:
Fixed.

pwittrock commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,231 @@ Documentation for other releases can be found at
+wordpress image includes an Apache server). We'll create two Kubernetes
+[Deployments](../../docs/user-guide/deployments.md) to run MySQL and
+WordPress, both with associated persistent volmes and
+[Services](../../docs/user-guide/services.md) to front each
```
is 'front' the right verb to use here?  expose?

jeffmendoza replied:
Reworded this whole section.

janetkuo commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -336,27 +266,33 @@ http://<external-ip>
 You should see the familiar WordPress init page.
 
+![WordPress init page](WordPress.png "WordPress init page")
+
```
Add the following warning message as we did in https://cloud.google.com/container-engine/docs/tutorials/persistent-disk/#visit_your_new_wordpress_blog?

> 
Warning: Do not leave your WordPress installation on this page. If it is found by another user, they can set up a website on your instance and use it to serve potentially malicious content. You should either continue with the installation past the point at which you create your username and password, delete your instance, or set up a firewall to restrict access.

jeffmendoza replied:
Added.

janetkuo commented on examples/mysql-wordpress-pd/mysql-deployment.yaml:
```diff
@@ -0,0 +1,61 @@
+    requests:
+      storage: 20Gi
+---
+apiVersion: v1
```
According to http://kubernetes.github.io/docs/user-guide/config-best-practices/, we should create a service before corresponding replication controllers so that the scheduler can spread the pods comprising the service.

Move service to the top?

jeffmendoza replied:
Fixed.

janetkuo commented on examples/mysql-wordpress-pd/README.md:
````diff
@@ -336,27 +266,33 @@ http://<external-ip>
-$ kubectl delete -f examples/mysql-wordpress-pd/wordpress.yaml
-$ kubectl delete -f examples/mysql-wordpress-pd/mysql.yaml
+```shell
+kubectl delete deployment,pod,service -l app=wordpress
````
Presumably `kubectl delete deployment` deletes the pods it controls cascadingly.

jeffmendoza replied:
Fixed, removed pod.

jeffmendoza commented:
I need to update the doc links

jeffmendoza commented:
Doc links updated.

Thanks reviewers, comments addressed.

amygdala commented on examples/mysql-wordpress-pd/README.md:
````diff
@@ -34,301 +34,246 @@ Documentation for other releases can be found at
-NAME                  CLUSTER_IP       EXTERNAL_IP       PORT(S)       SELECTOR               AGE
-wpfrontend            10.0.0.2         1.2.3.4           80/TCP        ...                    ...
-```
+The WordPress service has the setting `type: LoadBalancer`.  This will
````
Probably `type: LoadBalancer` still only works for some service providers? I wonder if we should add a note on that.

amygdala commented:
this is great, Jeff! thanks for doing this.

bgrant0607 commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,246 @@ Documentation for other releases can be found at
+
+Demonstrated Kubernetes Concepts:
+
+* [Persistent Volumes](https://github.com/kubernetes/kubernetes.github.io/blob/master/docs/user-guide/persistent-volumes.md) to
```
Today:
http://kubernetes.github.io/docs/user-guide/persistent-volumes/

Thursday:
http://kubernetes.io/docs/user-guide/persistent-volumes/

bgrant0607 replied:
(Everywhere that kubernetes.github.io is referenced)

bgrant0607 commented on examples/mysql-wordpress-pd/README.md:
````diff
@@ -34,301 +34,246 @@ Documentation for other releases can be found at
-```sh
-gcloud compute disks create --size=20GB --zone=$ZONE mysql-disk
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
````
Our doc munger can do this, also.

https://github.com/kubernetes/kubernetes/blob/master/docs/devel/how-to-doc.md#how-to-get-a-table-of-contents

jeffmendoza replied:
Ah, will do.

bgrant0607 commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,246 @@ Documentation for other releases can be found at
+* An [external load balancer](https://github.com/kubernetes/kubernetes.github.io/blob/master/docs/user-guide/services.md#type-loadbalancer)
+  will be used to access WordPress.
+* [Persistent Volume Claims](https://github.com/kubernetes/kubernetes.github.io/blob/master/docs/user-guide/persistent-volumes.md)
+  are used. You must create Persistent Volumes in your cluster to be
```
Is provisioning still alpha in 1.2? @saad-ali 
https://github.com/kubernetes/kubernetes/blob/master/examples/experimental/persistent-volume-provisioning/README.md

bgrant0607 commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,246 @@ Documentation for other releases can be found at
+Consult a
+[Getting Started Guide](https://github.com/kubernetes/kubernetes.github.io/blob/master/docs/getting-started-guides/README.md)
+to set up a cluster and the
+[kubectl](https://github.com/kubernetes/kubernetes.github.io/blob/master/docs/user-guide/prereqs.md) command line client.
```
command-line

bgrant0607 commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,246 @@ Documentation for other releases can be found at
+
+## Decide where you will store your data
+
+Both MySQL and WordPress will have a
```
MySQL and WordPress will each use a Persistent Volume to store their data.

bgrant0607 commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,246 @@ Documentation for other releases can be found at
+Both MySQL and WordPress will have a
+[Persistent Volume](https://github.com/kubernetes/kubernetes.github.io/blob/master/docs/user-guide/persistent-volumes.md) to
+store their data. We will use a Persistent Volume Claim to claim an
+available persistent volume. This example covers HostPath and
```
HostPath isn't used with Persistent Volumes, so this is confusing. Rephrase so that HostPath and Persistent Volume are 2 distinct cases.

bgrant0607 replied:
Nevermind, I guess hostPath is supported with PVs now.

bgrant0607 commented:
@jeffmendoza Please create an OWNERS file with your github id in it.

bgrant0607 commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,246 @@ Documentation for other releases can be found at
+Kubernetes cluster. The default setup script will create the cluster
+in the `us-central1-b` zone, as seen in the
+[config-default.sh](../../cluster/gce/config-default.sh) file. Replace
+`<zone>` below with the appropriate zone. The names `wordpress-1` and
```
Maybe `$ZONE` would be better than `<zone>`?

bgrant0607 replied:
I'm aware that we agreed on `<ZONE>` in the guidelines. :-/

bgrant0607 commented on examples/mysql-wordpress-pd/README.md:
````diff
@@ -34,301 +34,246 @@ Documentation for other releases can be found at
-
-```sh
-$ kubectl create -f examples/mysql-wordpress-pd/mysql-service.yaml
+Take a look at [mysql-deployment.yaml](mysql-deployment.yaml), and
````
You didn't want to auto-include the yaml?
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/how-to-doc.md#how-to-include-an-example

@johndmulhausen may also have a way to do that.

jeffmendoza replied:
If it were smaller, or if we could include a portion, yes. I'd rather not inline the whole thing though.

bgrant0607 commented on examples/mysql-wordpress-pd/wordpress-deployment.yaml:
```diff
@@ -0,0 +1,61 @@
+metadata:
+  name: wordpress
+  labels:
+    app: wordpress
```
In general, I'd put the same labels on the Service and PVC as on the corresponding Deployment.

jeffmendoza replied:
I have `app: wordpress` on all the items. On the pods I have `tier: frontend`, but not the deployment, claim, and service. Should I have `tier: frontend` on all those?

bgrant0607 commented on examples/mysql-wordpress-pd/mysql-deployment.yaml:
```diff
@@ -0,0 +1,61 @@
+  labels:
+    app: wordpress
+spec:
+  template:
```
Though updates aren't discussed in this example, the default update strategy wouldn't work, so I think we should add:
```
 strategy:
   type: Recreate
```
to both Deployments.

bgrant0607 commented on examples/mysql-wordpress-pd/README.md:
````diff
@@ -336,27 +281,40 @@ http://<external-ip>
 
-If you want to shut down the entire app installation, you can delete the services as well.
+```shell
+kubectl delete deployment,service -l app=wordpress
````
This also didn't delete the secret.

bgrant0607 commented:
Big improvement! Thanks!

pwittrock commented:
+1.  Definitely learned somethings from reading the example.

pwittrock commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,246 @@ Documentation for other releases can be found at
+  volumes, but any volume is sufficient.
+
+Consult a
+[Getting Started Guide](https://github.com/kubernetes/kubernetes.github.io/blob/master/docs/getting-started-guides/README.md)
```
This 404s now

pwittrock commented:
I took a fresh look at this today, and I think the setup section makes this example look hard.  Here are some of the things I think will be challenging:

- Cluster prerequisite section contains section about cluster requirements without telling the user how to setup the a cluster or how to verify this requirements.  The requirements link to detailed documents about each of the features, we shouldn't force the user down the path of reading these docs.  This section needs more context w.r.t. how to setup a cluster.  We should guide the user to a turn-key solutions that support the requirements, then only supply these as an 'if not using one of the k8s cluster providers below, first check that your k8s cluster supports x,y,z'

pwittrock commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,246 @@ Documentation for other releases can be found at
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+## Table of Contents
+
```
Is this the correct place for the table of contents?  Should it go before the tl;dr?  Before the what you will learn?  We should make the structuring consistent across examples.

jeffmendoza replied:
I like it here: Intro, tl;dr, then TOC.

People who want the tl;dr want to get there right away, TOC applies to the longer full sample.

pwittrock commented:
Overall I really like the way this example is looking.  Once finalized what do you think of pulling it out into a template for other examples?

Re: Cluster setup section
- Making bootstrapping a cluster easy is definitely out of the scope of this example
- As a stop gap we may want a simplified 'example cluster bootstrapping' page with fewer options, all of which have been tested to work on all examples, and a link to further reading.  Thoughts?
- I think each prerequisite should have a 1-liner command on how to verify it is met
- nit: How do you feel about putting the cluster setup line right before the prereqs instead of right after?  Can we ensure all the cluster setup options meet the prereqs by default?

pwittrock commented:
Should our examples have difficulty levels or be separated into beginner and more advanced sets?  The nfs example expects the user to be relatively familiar with k8s from the get go.

https://github.com/rootfs/kubernetes/blob/nfs-doc/examples/nfs/README.md

jeffmendoza commented:
Thanks, a template is a great idea.

As for the cluster setup. I think we want to stick to prereqs, and linking to the getting started guides. Here are the guidelines https://github.com/kubernetes/kubernetes/blob/master/examples/guidelines.md, though of course this is debatable. Specifically:

> * Should specify which release of Kubernetes is required and any other prerequisites, such as DNS, a cloudprovider with PV provisioning, a cloudprovider with external load balancers, etc.
>   * Point to general documentation about alternatives for those mechanisms rather than present the alternatives in each example.
>   * Tries to balance between using using new features, and being compatible across environments.

Let's focus on improving the docs that the prereqs link out to, and the getting started guides, this will improve all the samples as a whole

I don't think we can ensure all of the possible cluster setup options meet all the prereqs for all samples, as kubernetes is growing, adding more features, and being installed in new environments. This is why I changed it to have the prereqs first, so you know what you need before you go to the getting started guides to pick a deployment option. Honestly I think most people reading the sample already have an idea of where they want to run kubernetes if they are not doing so already.

Another thing to consider is that we're leaning towards examples being not-tutorials. In that they try to show how to run something, rather than teaching concepts through a story. So, I think it's ok for the examples to be closer to reference documentation over hand-holding.

I updated the examples readme https://github.com/kubernetes/kubernetes/blob/master/examples/README.md to have a chart with complexity level, Lets add to that as we decide which exmaples will be maintained.

jeffmendoza commented:
Forgot to mention, I like the idea of having a 1-liner command for each prereq, wherever it is possible. I'll add one for version, and look into the others. If we like it, I'll update the guidelines.

pwittrock commented:
SGTM.  Add a PTAL when ready for re-review.

pwittrock commented:
I tried running this on the local docker setup and got to:

```console
bash-3.2$ kubectl get services wordpress
NAME        CLUSTER-IP      EXTERNAL-IP   PORT(S)   AGE
wordpress   10.247.64.206                 80/TCP    11m
```

There is no external-ip.  Any idea what to do from here?

pwittrock commented:
I talked with @mikedanese about the local mac issues and filed #23093

jeffmendoza commented:
@pwittrock Updated PTAL

All of the ideas in #23093 sound really good. We need simple ways to query the cluster's capabilities, and some documentation on the various ways to access an external service.

pwittrock commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,249 @@ Documentation for other releases can be found at
+
+## Decide where you will store your data
+
+MySQL and WordPress will each use a
```
```
We will use a Persistent Volume Claim to claim an available persistent volume.
```

This reads awkwardly.  Maybe just embed this in the previous sentence to read 

MySQL and WordPress will each use a Persistent Volume with a Persistent Volume Claim to store their data.

pwittrock commented on examples/mysql-wordpress-pd/README.md:
````diff
@@ -34,301 +34,249 @@ Documentation for other releases can be found at
-        pdName: mysql-disk
-        fsType: ext4
-```
+This storage option is applicable if you are running on
````
s/applicable if/available on if/

pwittrock commented on examples/mysql-wordpress-pd/README.md:
```diff
@@ -34,301 +34,249 @@ Documentation for other releases can be found at
+set up the wordpress service behind an external IP.
 
-or by listing the forwarding rules for your project:
+Find the external IP for your WordPress service. It may take a minute
```
This is important and easy to miss if skimming the docs.  How do you feel about putting it in bold?

pwittrock commented:
Looking good.  Left a couple small comments - mainly apply bold that it takes a minute for the ip address to be assigned.

Send a PTAL when offline comments have been addressed.

pwittrock commented:
@jeffmendoza Don't want to hold his up any longer over some nits.  This is a huge improvement.  I can give this the lgtm if don't have any further changes.

jeffmendoza commented:
@pwittrock Updated with `tr --delete`, more cleanup instructions, and bold on the "please wait a minute".

pwittrock commented:
Looks great.
